### DISCUSSION TITLE
Deployment failed on varcel

### DISCUSSION THREAD

**Discussion by ibrahimkholil6210** [vercel/next.js#19648] [Help] [answered]
I trying to upload my NextJS project to varcel but the API routes couldn't be resolved while building the app what should i do.
![Capture](https://user-images.githubusercontent.com/33357137/100537356-0fe05800-3252-11eb-867c-eb874aaefdd8.PNG)

**Comment by waptik** [comment]
You can't use localhost url when deploying on vercel. It has to be a direct remote url.
Here's is how i do mine.

```js
// config.js
const app = {
  API_URL: process.env.API_URL ? process.env.API_URL : "https://localhost:4000",
  ...

};
```
Then i use it where i need it like this: `app.API_URL`.

In your case it should be:
```js
const API_URL = process.env.VERCEL_URL ? `${process.env.VERCEL_URL}/api : 'http://localhost:3000/api

// or if you have a something like `WEBSITE_URL` declared as environment variable

const API_URL = process.env.WEBSITE_URL ? `${process.env.WEBSITE_URL}/api : 'http://localhost:3000/api

```

**Reply by ibrahimkholil6210** [reply]
It is still failed with the same issue. Still trying to fetch data from localhost. Do I need to set any kind of flag before pushing for production?

**Reply by waptik** [reply]
I'm not so sure. You could try `fetch('/api/getAllproducts');` but i doubt it'll work.
I have this function that helps with dealing with issues like that.

```js
export const getBaseApiUrl = () => {
  consu url = process.browser ? "/api" :app.API_URL
  if (process.browser) {
    return "/api";
  }
// see `app` defined in my previous reply
  return url;
};

// usage inside getServerSideProps or getStaticProps

const products = await fetch(`${getBaseApiUrl()}/getAllproducts`)
...


```
So i defined `API_URL` in vercel dashboard to the full url of my remote api

**Comment by ibrahimkholil6210** [comment, marked as answer]
I solved the problem. As it turned out I was calling API routes from Data Fetching Functions which is forbidden. Letter i found it on documentation.
https://nextjs.org/docs/basic-features/data-fetching
Please read the Technical details section on this page for a better understanding.

**Reply by waptik** [reply]
Oh okay! I'm glad you were able to solve it on your own!
